sway-lib-std/src/hash.sw

```diff
@@ -124,3 +124,23 @@ pub fn hash_pair(value_a: b256, value_b: b256, method: HashMethod) -> b256 {
         }
     }
 }
+
+/// Returns the SHA-2-256 hash of `param`.
+pub fn sha256<T>(param: T) -> b256 {
+    let result_buffer: b256 = 0x0000000000000000000000000000000000000000000000000000000000000000;
+    if !is_reference_type::<T>() {
+        asm(r1: param, r2, r3: 8, r4: result_buffer) {
+            move r2 sp; // Move to register r2
+            cfei i8; // Grow stack by 1 word
+            sw r2 r1 i0; // Save value in register r1 to memory at r2
+            s256 r4 r2 r3; // Hash
+            r4: b256 // Return
+        }
```

Comment:
More hot tips for the aspiring ASM block author:
* The argument names don't have to resemble register names, they can be any identifier.  So you could use `asm(ptr: param, buf, eight: 8, hash: result_buffer) { ... }` and then the ASM would be slightly more readable.
* Although it isn't 100% necessary, I would really like to see a corresponding `CSEI` instruction after the hash is done.
* `// Move to register r2` is a bit misleading, as is `MOV` in general, since it's really just copying `$sp` to `r2`.  Just to be clear.

That second point might be contentious.  Taking some stack and then just forgetting about it is strictly leaking memory, which :triggered: me.  But OTOH, adding an extra instruction uses gas and the Fuel VM has enormous memory reserves to perhaps it's better just to leak.  @adlerjohn what do you say?

Hrmm, and this is absolutely assuming that structs and other aggregates have deterministically ordered elements and are packed tightly with no garbage bytes between elements.  This might be something we need to codify in the specs if we haven't already.  Or we make `T` belong to a `Serialized` trait which makes those guarantees for us.

Reply:
Yeah there's a balance to be struck between leaking memory (which is free so long as you don't run out of memory!), and reclaiming, which is strictly not-free (as in free beer). I'm leaning towards the standard library should be leak-free since it's expected to be used very heavily and there's a real possibility of running out of memory.

Reply:
> Hrmm, and this is absolutely assuming that structs and other aggregates have deterministically ordered elements and are packed tightly with no garbage bytes between elements. This might be something we need to codify in the specs if we haven't already. Or we make T belong to a Serialized trait which makes those guarantees for us.

Definitely something to work on in a follow-up PR, or several!